opamp/server_client.go

```diff
@@ -25,33 +30,210 @@ type serverClient struct {
+				s.logger.Error("Failed to connect to the server: %v", zap.Error(err))
+			},
+			OnErrorFunc: func(err *protobufs.ServerErrorResponse) {
+				s.logger.Error("Server returned an error response: %v", zap.String("", err.ErrorMessage))
+			},
+			GetEffectiveConfigFunc: func(ctx context.Context) (*protobufs.EffectiveConfig, error) {
+				cfg, err := s.configManager.CreateEffectiveConfigMsg()
+				if err != nil {
+					return nil, err
+				}
+				return cfg, nil
+			},
+			OnMessageFunc: s.onMessageFuncHandler,
+		},
+		Capabilities: protobufs.AgentCapabilities_AgentCapabilities_ReportsStatus |
+			protobufs.AgentCapabilities_AgentCapabilities_AcceptsRemoteConfig |
+			protobufs.AgentCapabilities_AgentCapabilities_ReportsRemoteConfig |
+			protobufs.AgentCapabilities_AgentCapabilities_ReportsEffectiveConfig |
+			protobufs.AgentCapabilities_AgentCapabilities_ReportsHealth,
+	}
+
+	err := s.opampClient.SetHealth(&protobufs.AgentHealth{Healthy: false})
+	if err != nil {
+		return err
+	}
+
+	err = s.opampClient.Start(ctx, settings)
+	if err != nil {
+		s.logger.Error("Error while starting opamp client", zap.Error(err))
+		return err
+	}
+	return s.collector.Run(ctx)
 }
 
 // Stop stops the Opamp client
 // It stops the Opamp client and disconnects from the Opamp server
 func (s *serverClient) Stop(ctx context.Context) error {
-	return nil
+	s.collector.Shutdown()
+	return s.opampClient.Stop(ctx)
+}
+
+func (s *serverClient) Error() error {
+	var err error
+	select {
+	case err = <-s.collector.ErrorChan():
+	default:
+	}
+	return err
 }
 
 // onMessageFuncHandler is the callback function that is called when the Opamp client receives a message from the Opamp server
 func (s *serverClient) onMessageFuncHandler(ctx context.Context, msg *types.MessageData) {
+	if msg.RemoteConfig != nil {
+		if err := s.onRemoteConfigHandler(ctx, msg.RemoteConfig); err != nil {
+			s.logger.Error("error while onRemoteConfigHandler", zap.Error(err))
+		}
+	}
+	// TODO: Handle other message types.
 }
 
 // onRemoteConfigHandler is the callback function that is called when the Opamp client receives a remote configuration from the Opamp server
 func (s *serverClient) onRemoteConfigHandler(ctx context.Context, remoteConfig *protobufs.AgentRemoteConfig) error {
+
+	changed, err := s.configManager.Apply(remoteConfig)
+	remoteCfgStatus := &protobufs.RemoteConfigStatus{
+		LastRemoteConfigHash: remoteConfig.GetConfigHash(),
+		Status:               protobufs.RemoteConfigStatuses_RemoteConfigStatuses_APPLIED,
+	}
+
+	if err != nil {
+		s.logger.Error("failed to apply config", zap.Error(err))
+
+		remoteCfgStatus.Status = protobufs.RemoteConfigStatuses_RemoteConfigStatuses_FAILED
+		remoteCfgStatus.ErrorMessage = fmt.Sprintf("failed to apply config changes: %s", err.Error())
+	}
+
+	if err := s.opampClient.SetRemoteConfigStatus(remoteCfgStatus); err != nil {
+		return fmt.Errorf("failed to set remote config status: %w", err)
+	}
+
+	if changed {
```

Comment:
It would have been good if opamp supported sending effective config along with remote config status. 

In the query service, we use effective config for any updates. If this calls fails, we might not have the current effective config on the server. Hence there can be mismatch. Though condition is an edge case. 

In query service side, may be we can sync pull instead of waiting for a push. Not an immediate need though.